src/github.com/getlantern/enproxy/conn_intf.go

```diff
@@ -200,60 +205,105 @@ func (c *Conn) Write(b []byte) (n int, err error) {
 }
 
 // Read() implements the function from net.Conn
-func (c *Conn) Read(b []byte) (n int, err error) {
+func (c *conn) Read(b []byte) (n int, err error) {
+	err = c.getAsyncErr()
+	if err != nil {
+		return
+	}
+
 	if c.submitRead(b) {
-		res, ok := <-c.readResponsesCh
-		if !ok {
-			return 0, io.EOF
-		} else {
-			return res.n, res.err
+		defer decrement(&blockedOnRead)
+
+		select {
+		case res, ok := <-c.readResponsesCh:
+			if !ok {
+				return 0, io.EOF
+			} else {
+				return res.n, res.err
+			}
+		case err := <-c.asyncErrCh:
+			return 0, err
 		}
 	} else {
 		return 0, io.EOF
 	}
 }
 
-// Close() implements the function from net.Conn
-func (c *Conn) Close() error {
-	c.closedMutex.Lock()
-	defer c.closedMutex.Unlock()
-	if !c.closed {
-		c.stopReadCh <- nil
-		c.stopWriteCh <- nil
-		c.stopRequestCh <- nil
-		c.closed = true
+func (c *conn) fail(err error) {
```

Comment:
We used to have some code paths that would fail and leave the connection hanging.  This new fail method allows them to report the error in a way that readers and writers can find out about it, and the connection is closed.